vortex-zigzag/src/compress.rs

```diff
@@ -30,87 +27,119 @@ impl EncodingCompression for ZigZagEncoding {
         parray
-            .stats()
-            .get_or_compute_cast::<i64>(&Stat::Min)
+            .statistics()
+            .compute_as_cast::<i64>(Stat::Min)
             .filter(|&min| min < 0)
             .map(|_| self as &dyn EncodingCompression)
     }
 
     fn compress(
         &self,
-        array: &dyn Array,
-        like: Option<&dyn Array>,
+        array: &Array,
+        like: Option<&Array>,
         ctx: CompressCtx,
-    ) -> VortexResult<ArrayRef> {
-        let zigzag_like = like.map(|like_arr| like_arr.as_zigzag());
-        let encoded = match ArrayKind::from(array) {
-            ArrayKind::Primitive(p) => zigzag_encode(p),
-            _ => unreachable!("This array kind should have been filtered out"),
-        }
-        .unwrap();
+    ) -> VortexResult<OwnedArray> {
+        let zigzag_like = like.map(|like_arr| like_arr.as_array_ref());
+        let encoded = if array.encoding().id() == Primitive::ID {
+            zigzag_encode(&array.as_primitive())?
+        } else {
+            println!("{}", array.encoding().id());
+            unreachable!("This array kind should have been filtered out")
+        };
 
-        Ok(
-            ZigZagArray::new(ctx.compress(encoded.encoded(), zigzag_like.map(|z| z.encoded()))?)
-                .into_array(),
-        )
+        Ok(OwnedZigZagArray::new(ctx.compress(&encoded.encoded(), zigzag_like)?).into_array())
     }
 }
 
-pub fn zigzag_encode(parray: &PrimitiveArray) -> VortexResult<ZigZagArray> {
+pub fn zigzag_encode(parray: &PrimitiveArray<'_>) -> VortexResult<OwnedZigZagArray> {
     let encoded = match parray.ptype() {
-        PType::I8 => zigzag_encode_primitive::<i8>(parray.buffer().typed_data(), parray.validity()),
+        PType::I8 => {
+            zigzag_encode_primitive::<i8>(parray.buffer().typed_data(), Some(parray.validity()))
+        }
         PType::I16 => {
-            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I32 => {
-            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I64 => {
-            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         _ => panic!("Unsupported ptype {}", parray.ptype()),
     };
-    ZigZagArray::try_new(encoded.into_array())
+    OwnedZigZagArray::try_new(encoded.into_array())
 }
 
-fn zigzag_encode_primitive<T: ZigZag + NativePType>(
-    values: &[T],
-    validity: Option<ValidityView>,
-) -> PrimitiveArray
+fn zigzag_encode_primitive<'a, T: ExternalZigZag + NativePType>(
+    values: &'a [T],
+    validity: Option<Validity<'a>>,
```

Comment:
Why is validity still optional?

Reply:
Wasn't sure if there was a good reason we wanted it to be optional, but given that the only usage involves wrapping a present validity in Some() I see your point.

Reply:
Before the port, Validity was optional. This is replaced by Validity::NonNullable